src/simsopt/solve/permanent_magnet_optimization.py

```diff
@@ -0,0 +1,415 @@
+        m = pm_opt.m0
+        for i in range(kwargs['max_iter_RS']):
+            # update m with the CONVEX part of the algorithm
+            algorithm_history, _, _, m = convex_step(
+                A_obj=pm_opt.A_obj,
+                b_obj=pm_opt.b_obj,
+                ATb=ATb,
+                m_proxy=np.ascontiguousarray(m_proxy.reshape(pm_opt.ndipoles, 3)),
+                m0=np.ascontiguousarray(m.reshape(pm_opt.ndipoles, 3)),  # note updated m is new guess
+                m_maxima=mmax,
+                **kwargs_convex
+            )
+            m_history.append(m)
+            m = np.ravel(m)
+            algorithm_history = algorithm_history[algorithm_history != 0]
+            errors.append(algorithm_history[-1])
+
+            # Solve the nonconvex optimization -- i.e. take a prox
+            m_proxy = prox(m, mmax, reg_rs, nu)
+            m_proxy_history.append(m_proxy)
+            if np.linalg.norm(m - m_proxy) < kwargs['epsilon_RS']:
+                print('Relax-and-split finished early, at iteration ', i)
+                break
+    else:
+        m0 = np.ascontiguousarray(m0.reshape(pm_opt.ndipoles, 3))
+        # no nonconvex terms being used, so just need one round of the
+        # convex algorithm called MwPGP
+        algorithm_history, _, m_history, m = convex_step(
+            A_obj=pm_opt.A_obj,
+            b_obj=pm_opt.b_obj,
+            ATb=ATb,
+            m_proxy=m0,
+            m0=m0,
+            m_maxima=mmax,
+            **kwargs_convex
+        )
+        m = np.ravel(m)
+        m_proxy = m
+
+    # note m = m_proxy if not using relax-and-split (i.e. problem is convex)
+    pm_opt.m = m
+    pm_opt.m_proxy = m_proxy
+    return errors, m_history, m_proxy_history
+
+
+def GPMO(pm_opt, algorithm='baseline', **algorithm_kwargs):
+    """
+    GPMO is a greedy algorithm alternative to
+    the relax-and-split algorithm for solving the permanent
+    magnet optimization problem. Full-strength magnets are placed
+    one-by-one according to minimize 
+    the MSE (fB). Allows for a number of keyword arguments that 
+    facilitate some basic backtracking (error correction) and 
+    placing magnets together so no isolated magnets occur.
+
+    Args:
+        pm_opt: PermanentMagnetGrid class object
+            The grid of permanent magnets to optimize.
+        algorithm: string
+            The type of greedy algorithm to use. Options are
+            baseline (the simple implementation of GPMO), multi
+            (place multiple magnets per iteration), backtracking
+            (backtrack every few hundred iterations to improve the
+            solution), ArbVec (the simple implementation of GPMO,
+            but with arbitrary oriented polarization vectors), and
+            ArbVec_backtracking (same as above but w/ backtracking).
+            Easiest to use is baseline but most effective algorithm
+            is ArbVec_backtracking.
+        algorithm_kwargs: dict
```

Comment:
Please use `kwargs`, which is the canonical python argument, in place of `algorithm_kwargs`.